### PR TITLE
FINERACT-1188: README and other files are not covered by Spotless

### PR DESCRIPTION
**Pull request by vidakovic**
## Description

This is a replacement for #1381 for FINERACT-1188

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

### CODE REVIEW

**Comment by vidakovic** [comment]
@ptuomola FYI: I even got the Eclipse launcher configuration fixed; when you import the project into your workspace it immediately starts a build. Bonus: the symbolic link to gradlew is also not necessary anymore (removed it).

**Comment by ptuomola** [comment]
@vidakovic looks like the integration tests are failing... did they work for you locally?

**Comment by ptuomola** [comment]
@vidakovic unfortunately this has broken Eclipse for me again - we are back at not finding Gradle:

Errors occurred during the build.
Errors running builder 'Integrated External Tool Builder' on project 'fineract-provider'.
Variable references non-existent resource : ${workspace_loc:/fineract/gradlew}
Variable references non-existent resource : ${workspace_loc:/fineract/gradlew}

If this works for you, are you setting up your Eclipse workspace based on the instructions in the README ie section "Instructions to run and debug in Eclipse IDE"? Or are you doing something else?

**Comment by vidakovic** [comment]
@ptuomola I have non of the errors you describe. Just to show you below a couple of screenshots from my Eclipse workspace:

... the error log shows one error entry, but related to the language server plugin; I think not related to anything we do:



... and this is the console output from the build that immediately starts after importing the project:



... the package explorer view looks also clean (no error indicators):

**Comment by vidakovic** [comment]
... (continued): I didn't run the integration tests locally (was a bit late already); what I did do is build the Docker image and run it; no problems there. I will checkout what happens with the integration tests; I see still no reason why they should fail. I was not aware of a specific Eclipse setup; will try to recreate your environment according to README instructions... maybe I find the issue like that. Will keep you posted.

Update: 

@ptuomola concerning Eclipse: I know now what's going on... I didn't execute "./gradlew cleanEclipse eclipse", because this is actually not necessary. If you import the project like this:



... then everything works perfectly well. From what I see we could get rid of the Eclipse Gradle plugin; even the Open JPA launcher configuration is automatically included in the project without any additional "./gradlew cleanEclipse eclipse". Unless there is another reason for this? Please let me know... I'll update the README if we can drop it.

**Comment by vidakovic** [comment]
Ok... one step ahead: finally have a candidate for the failing integration tests. I did an experiment with a locally running Fineract Docker container that had all database details setup correctly. I've disabled Carge in build.gradle temporarily and ran the tests again... voila it worked. I think what happens with Cargo: it doesn't find the deployable WAR file; it doesn't seem that Fineract is started with Cargo at all.

**Comment by vidakovic** [comment]
@ptuomola yup, it was the Cargo plugin. The problem was that for some reason it didn't detect the WAR file anymore... so I added a deployable section. Just ran the integration tests locally, no failures. Let's see how the Travis build goes to confirm. If it's successful would be great if you could have another look at this. Please let me know if my notes on Eclipse are OK for you; the "new" instructions are actually shorter than the current README and from what I see works at least as well as before.

**Comment by ptuomola** [comment]
@vidakovic Ah I see - you're importing this as an Gradle project rather than Eclipse project. 

But if you do that, do you get any of the benefits of Eclipse - e.g. debugging, hot code replace, automatic incremental recompile etc?

**Comment by vidakovic** [comment]
@ptuomola as far as I can see all the same

**Comment by ptuomola** [comment]
@vidakovic cool - just tried this and yes, it seems to work equally well. So no issue from Eclipse perspective.

But the build still seems to be failing? Think there's some problem with the deployment to Cargo:
> No such property: war for class: java.lang.String

**Comment by vidakovic** [comment]
@ptuomola this Cargo plugin is a gift that just keeps giving... let me see

**Comment by vidakovic** [comment]
Problem was the use of "." right after a variable in a Groovy string; fixed

**Comment by vidakovic** [comment]
... of course there's something else... but it's something in the Travis configuration... all the tests etc. went through without a problem. Looking into this.

File: .travis.yml
Comment (ptuomola) [inline .travis.yml:75]: Hmm... does it make sense that the generated code for client is within fineract-provider? Would it not make more sense to generate it into fineract-client or something? Or is that the next step in the process?

File: build.gradle
Comment (ptuomola) [inline build.gradle:35]: I think the other OpenJPA version is now in fineract-provider/build.gradle rather than "below"

File: fineract-provider/build.gradle
Comment (ptuomola) [inline fineract-provider/build.gradle:35]: Should we also remove the eclipse plugin as it doesn't work anymore (as discussed in the JIRA)?

File: fineract-provider/build.gradle
Comment (ptuomola) [inline fineract-provider/build.gradle:580]: As per JIRA-1187 ticket, we should keep the name of the war file as "fineract-provider.war" without the release version. Otherwise this will confuse everyone who tries to deploy it to Tomcat and finds that the URL to access Fineract keeps changing every time there's a new version...

**Review by ptuomola — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (vidakovic) [reply .travis.yml:75]: It's on the list... the next step ;-)

File: build.gradle
Comment (vidakovic) [reply build.gradle:35]: I'm not sure if I get this one? What's the problem here?

File: fineract-provider/build.gradle
Comment (vidakovic) [reply fineract-provider/build.gradle:35]: Yes, I was just waiting for you to confirm. Will do this right away.

File: fineract-provider/build.gradle
Comment (vidakovic) [reply fineract-provider/build.gradle:580]: Can we take care of that separately? Because this is already in place upstream and it would force me to touch the distribution configuration and the release configuration... which would be more than needed to make this work here. But I can take care of this once we are done with multi-module (or squeeze it in between with a separate PR). Good for you?

**Comment by vidakovic** [comment]
@ptuomola please see my replies; I removed the Eclipse Gradle plugin references and updated the README

File: fineract-provider/build.gradle
Comment (ptuomola) [reply fineract-provider/build.gradle:580]: Hmm... not sure I understand. After FINERACT-1187 was merged 4 days ago, we were building fineract-provider.war. Are you saying that some other change has now broken that again, and we are back to building fineract-provider-releaseVersion.war?

Comment (vorburger) [reply fineract-provider/build.gradle:580]: But wouldn't that be a "regression" of FINERACT-1187? Why do we have to break that here?

File: build.gradle
Comment (ptuomola) [inline build.gradle]:
```suggestion
        classpath 'org.apache.openjpa:openjpa:3.1.2' // when upgrading, also change OpenJPA version repeated in fineract-provider/build.gradle!
```

File: build.gradle
Comment (ptuomola) [reply build.gradle:35]: Sorry got the wrong line - added the corrected one above

File: fineract-provider/build.gradle
Comment (vidakovic) [reply fineract-provider/build.gradle:580]: I was not aware of that change... that's all. If you say that this is already working in develop then I have to pull in the recent changes; that's all.

Is FINERACT-1187 already merged? I just checked upstream... there are no changes that anyone took care of FINERACT-1187... will do this now... just to get this thing over the finish line.

File: build.gradle
Comment (ptuomola) [inline build.gradle]: This line is resulting in regression of FINERACT-1187... i.e. when you define a property "version" then war task will automatically append the "version" to the WAR name

File: fineract-provider/build.gradle
Comment (ptuomola) [reply fineract-provider/build.gradle:580]: FINERACT-1187 was merged and working. It's change to build.gradle which is part of this PR that is breaking it again. I've marked the problematic line above with a review comment.

Comment (ptuomola) [reply fineract-provider/build.gradle:580]: And of course once you fix that problem in build.gradle, then this change to cargo is not needed...

Comment (vorburger) [reply fineract-provider/build.gradle:580]: To know if fixes are merged, check that Links To section thingie on the FINERACT-1187 JIRA... notice the link to #1379 - makes sense? In general, any issues in Status Resolved,  Resolution: Fixed should have associated Links To PRs... if they don't, we did something wrong.

File: build.gradle
Comment (vidakovic) [reply build.gradle]: Fixed and pushed. 

Suggestion for another time: I think this would be easier by defining a variable in buildscript ext section and use that in both places. Then we can get rid of these comments.

File: build.gradle
Comment (vidakovic) [reply build.gradle]: Alright... reverted that. Everything pushed.

**Comment by vidakovic** [comment]
@ptuomola Spotless complains a format issue in README line 200, but when I run locally it doesn't. Any ideas? Cannot reproduce. Could this be coming from another commit upstream?

Update: I've merged the upstream changes.  The interesting part was only the new ActiveMQ section in the README... which had a forbidden space at the end of the title.

**Comment by vidakovic** [comment]
@vorburger @ptuomola wow... please let's merge before something else happens upstream ;-)

**Comment by vorburger** [comment]
@vidakovic Whoa! Nice. I'll review ASAP, but 1 quick minor practical feedback already: If you change the usual merge drop-down (above) to the _Rebase and merge_ which we typically prefer using (see https://github.com/apache/fineract#merge-strategy), you'll see that GitHub shows _This branch cannot be rebased due to conflicts_. That means we would HAVE to merge your 2 commits here with _Create a merge commit_, which is possible, but always a bit "ugly" and we try to avoid, or with _Squash and merge_ and that creates weird author/ownership metadata (check out e.g. latest e57b7da). If you don't mind, just do a `git rebase develop` (fixing anything required), and `git push --force  vidakovic`, and it will be a little bit "cleaner". With that out of the way, I'll review the actual changes now! :smiling_imp:

File: README.md
Comment (vorburger) [inline README.md:87]: @vidakovic thanks for this updae, @ptuomola  FYI this is actually how I use it as well (without `./gradlew cleanEclipse eclipse`)

File: build.gradle
Comment (vorburger) [inline build.gradle:56]: FTR, the `ext['groovy.version']` in `fineract-provider/build.gradle` used to be 3.0.3, and this upgrades it to `3.0.6`. I'm sure that's perfectly fine, so more as a general comment, when we move things around like this, it's best to move "as is", and then make changes (upgrades, here) separately subsequently (or before, if required). I'm not suggesting that you change it here anymore, just commenting in general, for future refactorings (of build or code, same).

Comment (vidakovic) [reply build.gradle:56]: ok

File: fineract-provider/build.gradle
Comment (vorburger) [inline fineract-provider/build.gradle:67]: Right, we can ditch this.. makes sense, to me. Again, better separately in the future, but not suggesting to touch here anymore (sorry to keep being a pain about this).

File: fineract-provider/build.gradle
Comment (vorburger) [inline fineract-provider/build.gradle:185]: so this gets entirely ditch, not just moved.. why? @ptuomola does this LGTY? (I have NOT pulled it myself yet, to actually test.)

File: build.gradle
Comment (vorburger) [inline build.gradle:88]: Nice!!

File: build.gradle
Comment (vorburger) [inline build.gradle:126]: This seems to be moving from another section? More out of curiosity, what is this for? /Cc @thesmallstar (FYI we are doing a massive "move", see FINERACT-1188 and FINERACT-1171 for more background)

**Comment by vidakovic** [comment]
@vorburger could you give me the instructions again to solve that issue you mentioned with the rebase? I tried the steps you suggested, but just get weird feedback on the console. E. g. should I do "git rebase develop" in the PR branch? My repos develop and the PR branch are already in sync.

File: fineract-provider/build.gradle
Comment (vidakovic) [reply fineract-provider/build.gradle:185]: ... because "fineract-provider/.externalToolBuilders/OpenJPA Enhance Builder.launch" gets picked up automatically by Eclipse when you import via "Existing Gradle Project"... when I open the .project file in fineract-provider then I see this:



So... should be fine.

File: fineract-provider/gradlew
Comment (vorburger) [inline fineract-provider/gradlew:1]: FTR: It took me a sec to understand that you are removing the symlink here.. doesn't work? But probably a good idea anyway, for Windoze consistency.. I noticed that currently on develop we have a fineract-provider/gradlew (symlink), but no fineract-provider/gradlew.bat - that would probably create confusion, so +1 to rm this.

**Comment by vorburger** [comment]
> @vorburger could you give me the instructions again to solve that issue you mentioned with the rebase? 

    git rebase develop 
    git push --force vidakovic

> I tried the steps you suggested, but just get weird feedback on the console. 

do you want to share the git error message / feedback?

> E. g. should I do "git rebase develop" in the PR branch? 

Yeah.

> My repos develop and the PR branch are already in sync.

How do you mean? This PR makes it look like you are working on `feature/FINERACT-1188-fix` branch... if you are saying that your local `develop` branch already includes your `feature/FINERACT-1188-fix` commit, then my recommendation would have to be to avoid that. It's confusing as well. You COULD use `git rebase origin/develop` instead of `git rebase develop`, to tell it that you want "the _develop_ branch from the _origin_ remote" (assuming that `git remote -v` shows you that your `origin` is `git@github.com:apache/fineract.git`, but you SHOULD not have to do that.. it's super confusing if you locally make `develop` divert from `origin/develop` (other than it sometimes behind "behind, but fast-forwardable", that's normal). So, long story short, if I were you, I would `git checkout develop && git reset --hard origin/develop` - BUT be careful, you'll LOOSE what you have on your `develop` and don't already have anywhere else, so be sure all this makes sense! (Just ask if not; happy to elaborate, but tomorrow.)

**Review by vorburger — APPROVED** [review]
LGTM! I even pulled and "tested" it, and indeed, lo and behold, if I add a space at the end of a line in README.md and run `./gradlew spotlessCheck`, it chokes on it...  works as advertised (i.e. implements FINERACT-1188 as described). I'll let @ptuomola review and Approve this as well. If any other committers are reading along here, feel free to Approve or Comment as well, of course.

If you can sort out the `rebase` so this has no merge commit, it would be neat. I'll help you, if required.

**Comment by vidakovic** [comment]
@vorburger I think I get an idea what happened here with the rebase... I first tried to solve this in the browser, but somehow lost my changes... and then "fixed" the conflict by merging the PR branch into my vidakovic/fineract develop branch (which was suggested by Github... don't remember where I read)... and that lead to the current PR branch.

So, I'll try to bring my develop branch again to the same level as upstream - without my additional conflict fix merge. And then your instructions should work. Will let you know here... thanks again.

Update: sweet... that worked... learned again something useful!

**Review by ptuomola — APPROVED** [review]
LGTM! Thanks for persevering with this.

I can merge this - unless someone else wants to review first?

**Comment by vidakovic** [comment]
Thanks again for the patience ☺️ @ptuomola and @vorburger ... now that we are all happy with it I guess you can merge whenever you have a moment

**Pull request merged**
